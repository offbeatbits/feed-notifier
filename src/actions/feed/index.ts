export * from './fetch';
export * from './getLatestArticle';
