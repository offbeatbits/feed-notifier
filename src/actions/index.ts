import { update } from '../utils';

export {
  update,
};
