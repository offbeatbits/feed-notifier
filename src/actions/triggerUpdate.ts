import { logger } from '../lib';

import {
  channels,
  clients,
  feedUrl,
} from '../config';

import {
  updateAllChannels,
  updateSingleChannel,
} from './update';

import {
  fetch as fetchFeed,
} from './feed';

export const triggerUpdate = async () => {
  const content = 'foo';
  logger.verbose(`Found channel feed URL: ${feedUrl}`);

  logger.verbose('Fetching feed...');

  const feed = await fetchFeed(feedUrl);

  console.log(feed.getFeedUpdateDate()); // eslint-disable-line
  console.log(feed.getLatestPostDate()); // eslint-disable-line
  console.log(feed.getLatestPostTitle()); // eslint-disable-line
  console.log(feed.getLatestPostUrl()); // eslint-disable-line
  console.log(feed.getLatestPostId()); // eslint-disable-line

  logger.info('Updating all channels...');

  updateAllChannels({
    channels,
    clients,
    updateFn: updateSingleChannel,
    content,
  });

  logger.info('Update complete!');
};
