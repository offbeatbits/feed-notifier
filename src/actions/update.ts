import {
  channels,
  clients,
} from '../config';
import {
  createService,
  getChannelConfigArray,
} from '../helpers';
import {
  logger,
} from '../lib';
import { ChannelConfigRecord } from '../types';

/* eslint-disable no-console */

const updateSingleChannel = ({ type, settings }: ChannelConfigRecord) => {
  logger.info(`Updating channel '${type}'...`);
  if (settings.enabled) {
    const service = createService(type, settings);
    (service as any).testMethod();
    logger.info(`Channel '${type}' updated!`);
  } else {
    logger.info(`Channel '${type}' disabled. Skipping...`);
  }
};

export const updateAllChannels = () => {
  logger.info('Updating all channels...');
  const channelsToUpdate = getChannelConfigArray(channels, clients);
  channelsToUpdate.forEach(channel => updateSingleChannel(channel));
  logger.info('Update complete!');
};
