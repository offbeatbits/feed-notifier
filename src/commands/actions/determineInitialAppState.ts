import { ArticleObject } from "../../types";
import { logger } from "../../lib";
import { isItForcedUpdate, isItInitialization } from "../../helpers";

export const determineInitialAppState = (
  cachedArticle: ArticleObject | null
) => {
  logger.info("Update process started!");
  const isForcedUpdate = isItForcedUpdate();
  const isInitialization = isItInitialization();
  if (!cachedArticle) logger.verbose("Article cache is empty!");

  if (isForcedUpdate && isInitialization) {
    logger.verbose(
      "Detected '--skipCache' and '--initialize' passed at once. Make up your mind, please? :)"
    );
    return false;
  }
  if (!cachedArticle && !isForcedUpdate && !isInitialization) {
    logger.verbose(
      "Exiting (pass '--skipCache' to force an update or '--initialize' to generate cache)"
    );
    return false;
  }
  if (isForcedUpdate) {
    logger.verbose("Proceeding despite empty cache...");
    return true;
  }
  return true;
};
