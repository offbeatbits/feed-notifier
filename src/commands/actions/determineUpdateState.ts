import { ArticleObject } from "../../types.d";
import { logger } from "../../lib";
import { areArticlesDifferent, isItForcedUpdate } from "../../helpers";

export const determineUpdateState = (
  oldArticle: ArticleObject | null,
  newArticle: ArticleObject | null
) => {
  const skip = isItForcedUpdate();

  const isUpdateNeeded = areArticlesDifferent({
    oldArticle,
    newArticle,
    skip,
  });

  logger.info(`Update needed: ${isUpdateNeeded}`);
  return isUpdateNeeded;
};
