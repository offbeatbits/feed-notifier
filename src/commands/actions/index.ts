export * from './determineInitialAppState';
export * from './determineUpdateState';
export * from './prepareUpdate';
export * from './postTheUpdate';
export * from './updateCache';
export * from './finishUpdate';
