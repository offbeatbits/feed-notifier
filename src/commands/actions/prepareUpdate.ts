import { ArticleObject } from "../../types.d";
import { logger } from "../../lib";
import { generatePostContent } from "../../helpers";

export const prepareUpdate = (latestArticle: ArticleObject) => {
  logger.info("New article detected - generating update content...");
  return generatePostContent(latestArticle);
};
