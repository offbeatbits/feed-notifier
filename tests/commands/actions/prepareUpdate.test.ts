import { prepareUpdate } from '../../../src/commands/actions/prepareUpdate';

describe('prepareUpdate', () => {
  const article = {
    feedUpdateDate: '2021-06-04T11:20:42Z',
    date: '2021-06-04T11:20:42Z',
    title: 'prepared_sample_title_1',
    url: 'prepared_sample_url_1',
  };

  it('should prepare new article for update', () => {
    expect(prepareUpdate(article)).toMatchSnapshot();
  });
});
