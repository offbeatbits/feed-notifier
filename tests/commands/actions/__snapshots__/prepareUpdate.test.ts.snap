// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`prepareUpdate should prepare new article for update 1`] = `"prepared_sample_title_1 prepared_sample_url_1"`;
